src/tests/test_dilithium.cpp

```diff
@@ -0,0 +1,295 @@
+					if (V[j] == 0xff)
+						V[j] = 0x00;
+					else {
+						V[j]++;
+						break;
+					}
+				}
+
+				AES256_ECB(Key, V, temp + 16 * i);
+			}
+			if (provided_data != NULL)
+				for (int i = 0; i < 48; i++)
+					temp[i] ^= provided_data[i];
+			memcpy(Key, temp, 32);
+			memcpy(V, temp + 32, 16);
+		}
+
+			void clear() override
+			{
+				// reset struct
+				memset(DRBG_ctx.Key, 0x00, 32);
+				memset(DRBG_ctx.V, 0x00, 16);
+				DRBG_ctx.reseed_counter = 0;
+			}
+
+			bool accepts_input() const override { return true; }
+
+			void add_entropy(const uint8_t data[], size_t len) override
+			{
+				BOTAN_UNUSED(len);
+				randombytes_init(data, nullptr, 256);
+			}
+
+			bool is_seeded() const override
+			{
+				return true;
+			}
+
+			void randomize(uint8_t out[], size_t len) override
+			{
+				randombytes(out, len);
+			}
+
+			Dilithium_Test_RNG(const std::vector<uint8_t>& seed)
+			{
+				clear();
+				add_entropy(seed.data(), seed.size());
+			}
+
+		private:
+			void randombytes_init(const unsigned char* entropy_input, unsigned char* personalization_string, int security_strength)
+			{
+				BOTAN_UNUSED(security_strength);
+				unsigned char   seed_material[48];
+
+				memcpy(seed_material, entropy_input, 48);
+				if (personalization_string)
+					for (int i = 0; i < 48; i++)
+						seed_material[i] ^= personalization_string[i];
+				memset(DRBG_ctx.Key, 0x00, 32);
+				memset(DRBG_ctx.V, 0x00, 16);
+				AES256_CTR_DRBG_Update(seed_material, DRBG_ctx.Key, DRBG_ctx.V);
+				DRBG_ctx.reseed_counter = 1;
+			}
+
+			int randombytes(unsigned char* x, size_t xlen)
+			{
+				unsigned char   block[16];
+				int             i = 0;
+
+				while (xlen > 0) {
+					//increment V
+					for (int j = 15; j >= 0; j--) {
+						if (DRBG_ctx.V[j] == 0xff)
+							DRBG_ctx.V[j] = 0x00;
+						else {
+							DRBG_ctx.V[j]++;
+							break;
+						}
+					}
+					AES256_ECB(DRBG_ctx.Key, DRBG_ctx.V, block);
+					if (xlen > 15) {
+						memcpy(x + i, block, 16);
+						i += 16;
+						xlen -= 16;
+					}
+					else {
+						memcpy(x + i, block, xlen);
+						xlen = 0;
+					}
+				}
+				AES256_CTR_DRBG_Update(NULL, DRBG_ctx.Key, DRBG_ctx.V);
+				DRBG_ctx.reseed_counter++;
+
+				return 0;
+			}
+
+			typedef struct {
+				unsigned char   Key[32];
+				unsigned char   V[16];
+				int             reseed_counter;
+			} AES256_CTR_DRBG_struct;
+			AES256_CTR_DRBG_struct  DRBG_ctx;
+		};
+
+		Test::Result run_dilithium_kat_test(const char* test_name, const VarMap& vars, Botan::DilithiumDimension dimension, Botan::DilithiumAlgo algo,
+			[[maybe_unused]] const std::string& algo_name)
+		{
+			Test::Result result(test_name);
+			int                 ret_val;
+
+			// read input from test file
+			const auto ref_seed = vars.get_req_bin("seed");
+			const auto ref_mlen = vars.get_req_sz("mlen");
+			const auto ref_msg = vars.get_req_bin("msg");
+			const auto ref_pk = vars.get_req_bin("pk");
+			const auto ref_sk = vars.get_req_bin("sk");
```

Comment:
The secret key is large and takes up a lot of space in the files. Instead include only the seed, and re-derive the public and private keys. This has the nice side effect of testing key generation quite thoroughly.

Reply:
Oh, I see we are already doing rederivation - great! So in that case replace the secret and public keys included in the test data files with the SHA-3 of their contents, so we can still check that the generated key is correct. (I suggest SHA-3 simply because Dilithium requires it already)